flink-core/src/main/java/org/apache/flink/core/memory/MemorySegment.java

```diff
@@ -217,10 +222,14 @@ public int size() {
     /**
      * Checks whether the memory segment was freed.
      *
+     * <p>This method internally involves cross-thread synchronization. Do not use for performance
+     * sensitive code paths.
+     *
      * @return <tt>true</tt>, if the memory segment has been freed, <tt>false</tt> otherwise.
      */
     public boolean isFreed() {
-        return address > addressLimit;
+        // in performance sensitive cases, use 'address > addressLimit' instead
+        return isFreedAtomic.get();
```

Comment:
I think we could promote this atomic handling to `MemoryUtils.createMemoryCleaner` or even `MemoryUtils.allocateUnsafe`(with enriched struct). Then we could remove this caveat. I think we are only guarding unsafe part ? If yes, I recommend to move this concurrent guarding to unsafe`cleaner`.

Reply:
I guess that's a trade-off. The more operations we guard against concurrency, the more affected by cross-thread synchronization. In this case, I'm in favor of guarding all the segment freeings, rather than only the unsafe cleaners. The performance cost on calling `free()` should be acceptable given that each segment should only be freed once.

As for `isFree()`, it's currently only used for checking before releasing the segment, or in test cases. I'm adding this warning message trying to prevent it from being used in performance sensitive cases in future, like in `put()` and `get()`. Alternatively, I'm also ok with not changing this method, living with the fact that freeing of the segment may not always be visible to this method immediately. In that case, we may want to add a warning on that as well.

Reply:
Taking possible concurrent accessing into account, `isFree()` is always best effort that says later access is not guaranteed to success. I pointed this out in [#15246](https://github.com/apache/flink/pull/15246#discussion_r595425459). I suggest to make it private and keep it untouched.

For whether to guard all segments freeings, I'd prefer to only unsafe part, but it is ok to go other direction. Here is my considerations:
* `MemorySegment.free` is not only used by `MemoryManager` but also `FreeingBufferRecycler` which is spread over network stack. In there, a volatile access is counted-:). See `AvailabilityProvider.isAvailable`.
* For safe part, it is far less destructive comparing to unsafe part. I think this "destructive" motivate us for this protection, at least for me.

All for all, we are guarding freeing not accessing.

Reply:
I agree that `isFree()` does not guarantee success of later access anyway. I'd keep it untouched and make it package private (for testing purpose).

As for the atomic handling, I'd prefer to keep it as is. Network stack is performance sensitive because many of the operations are on per-record code paths. However, as far as I checked, freeing of memory segments are not on per-record paths. (And if they are, that probably indicates `FreeingBufferRecycler` should not be used.)

Moreover, once we have hunted down all the multiple-free cases, we can enforce a strict failing against multiple-frees. At that time, we won't need this atomic checking anymore. Therefore, I'm in favor of not introduce the complexity of an enriched cleaner struct, which will soon be removed.

Reply:
For `isFree()`: Sorry for misleading. `VisibleForTesting` is sufficient.

For atomic handling placement: I am ok to either approaches. But I still think it is a good to not spread this concurrent guarding to global.

Besides above, I am a bit curious about how to do leak detection in current direction: `finalize` in `MemorySegment` ? Seems to me that, all these changes might affect streaming users without much benefits. The problems are two folds:
1. `finalize` itself is a yes or no in coding phase.
2. `finalize` is applied to all `MemorySegment` no matter whether it is need or not or active-freed.

It is a bit out of context and should belong to FLINK-21799. But I think it is related as we might run out of choices if opt out gc-cleaner approaches and keep only raw address directly in `MemorySegment`.

> Moreover, once we have hunted down all the multiple-free cases, we can enforce a strict failing against multiple-frees. At that time, we won't need this atomic checking anymore. Therefore, I'm in favor of not introduce the complexity of an enriched cleaner struct, which will soon be removed.

Personally, I would prefer to keep following in long term.
* Keep concurrent guarding for unsafe memory due to bad experience for crash and investigation.
* Configurable leak detection. Options are "none", "log" and "fail". This is a redraw from my side in FLINK-21419. `finalize` does not meet.

Reply:
In long term, I think it would be better to rule out all of the following cases. And by "rule out", I mean getting rid of the existing cases and prevent new cases from being introduced.
- A segment is freed more than once (concurrently or not)
- A segment is being GC-ed without having been freed

Once the existing cases are all removed, we can strictly fail on these cases. That also means guarding concurrent frees should not be needed. Any concurrent frees should have already failed due to multiple frees during CI tests.

The benefit of enforcing these rules on memory types that it simplifies the contract of using a memory segment and improves maintainability. Otherwise, it would be complicated to understand that a segment has to be freed exactly once if its unsafe, while it's okay to free another segment multiple times (or never) if it's heap/direct. Users of `MemorySegment` should not be forced to understand the underlying memory type.

I admit enforcing such rules comes at a price of extra overhead in freeing/finalizing the segments. However, I'd argue that the price is insignificant compared to the benefits.

Reply:
I have no objection for enforcing rules on memory types. The visible difference is minor: concurrent free will fail for all memory segments.

I guess it is price of full aligned api unification for safe and unsafe memory segment. I think we could go through current approach and see whether it introduces much overhead for streaming users. In later stage, we could decide whether to drop the whole things and compromise a bit in implementation or api semantics base on concrete cases/usages/benchmarks.

Though, I still think it is a good to guard unsafe part even after all existing invalid cases detected especially it is uncertain(FLINK-21798) that unsafe part could be "unmanaged". It is a safety net for that uncertain future, otherwise we are just rollback ~ rollback ~ rollback things. That is why I think configurable is good.

Reply:
I think we have both made a point.
1. A memory segment should deal with multiple-frees, concurrency, leaks regardless of memory type.
2. The unsafe memory (de-)allocation should be self-contained, without relying on the correctness of memory segment and its usages. To be specific, the cleaner itself should make sure it is called exactly once.

I'm leaning towards `1` for its broader coverage, and is neutral to `2`. I think `2` might not be necessary if `1` works properly, but it does not harm to have such an extra safe net given that as far as I can see the price is high.

Reply:
I guess so. I believed that `1` covers `2` and `1` will hunt more cases. What I am afraid of is: we will rollback to old situation after all existing cases hunted and `1` ruled out.

I guess we could also reconsider this if unsafe part is used as "unmanaged" in future. I hope they are always "managed".